pkg/endpoint/bpf.go

```diff
@@ -212,76 +218,79 @@ func (e *Endpoint) addNewRedirectsFromDesiredPolicy(ingress bool, desiredRedirec
-			// is HTTP, just allow traffic to the port at L4 by setting the proxy port
-			// to 0.
-			if !e.hasSidecarProxy || l4.L7Parser != policy.ParserTypeHTTP {
-				var finalizeFunc revert.FinalizeFunc
-				var revertFunc revert.RevertFunc
-
-				pp := e.newProxyPolicy(l4, dstPort)
-
-				var err error
-				redirectPort, err, finalizeFunc, revertFunc = e.proxy.CreateOrUpdateRedirect(e.aliveCtx, &pp, proxyID, e, proxyWaitGroup)
-				if err != nil {
-					// Skip redirects that can not be created or updated.  This
-					// can happen when a listener is missing, for example when
-					// restarting and k8s delivers the CNP before the related
-					// CEC.
-					// Policy is regenerated when listeners are added or removed
-					// to fix this condition when the listener is available.
-					e.getLogger().WithField(logfields.Listener, l4.GetListener()).WithError(err).Debug("Redirect rule with missing listener skipped, will be applied once the listener is available")
-					return 0, false
+		func(l4 *policy.L4Filter) bool {
+			for _, v := range l4.PerSelectorPolicies {
+				if v == nil || !v.IsRedirect() {
+					continue
+				}
+				// proxyID() returns also the destination port for the policy,
+				// which may be resolved from a named port
+				proxyID, dstPort := e.proxyID(l4, v.Listener)
+				if proxyID == "" {
+					// Skip redirects for which a proxyID cannot be created.
+					// This may happen due to the named port mapping not
+					// existing or multiple PODs defining the same port name
+					// with different port values. The redirect will be created
+					// when the mapping is available or when the port name
+					// conflicts have been resolved in POD specs.
+					continue
+				}
+				// desiredRedirects starts out empty, so we can use it check
+				// if the redirect has already been updated on this round.
+				if desiredRedirects[proxyID] {
+					continue
 				}
-				finalizeList.Append(finalizeFunc)
-				revertStack.Push(revertFunc)
-			}
 
-			if e.realizedRedirects == nil {
-				e.realizedRedirects = make(map[string]uint16)
-			}
-			if _, found := e.realizedRedirects[proxyID]; !found {
-				revertStack.Push(func() error {
-					delete(e.realizedRedirects, proxyID)
-					return nil
-				})
+				var redirectPort uint16
+
+				// Only create a redirect if the proxy is NOT running in a sidecar container
+				// with an HTTP parser. If running in a sidecar container and the parser
+				// is HTTP, just allow traffic to the port at L4 by setting the proxy port
+				// to 0.
+				if !e.hasSidecarProxy || l4.L7Parser != policy.ParserTypeHTTP {
+
+					pp := e.newProxyPolicy(l4, v, dstPort)
+					proxyPort, err, finalizeFunc, revertFunc := e.proxy.CreateOrUpdateRedirect(e.aliveCtx, &pp, proxyID, e, proxyWaitGroup)
+					if err != nil {
+						// Skip redirects that can not be created or updated.  This
+						// can happen when a listener is missing, for example when
+						// restarting and k8s delivers the CNP before the related
```

Comment:
Yes, packets hitting policies with listener reference can get dropped due to listener not existing yet. Any CEC/CCEC ingestion causes the endpoint policies to be regenerated so that any drops due to this are only momentary.